src/features/weak-references.md

```diff
@@ -54,104 +54,133 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
+
+Until now, the solution is to manually unregister the event listener via `dispose` method.
 
 ```js
-const cache = new Map();
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    this.socket = socket;
+    this.listener = (ev) => { this.events.push(ev); };
+    socket.addEventListener('message', this.listener);
+  }
 
-function getImageCached(name) {
-  const ref = cache.get(name);
-  if (ref !== undefined) {
-    const deref = ref.deref();
-    if (deref !== undefined) return deref;
+  dispose() {
+    this.socket.removeEventListener('message', this.listener);
   }
-  const image = performExpensiveOperation(name);
-  const wr = new WeakRef(image);
-  cache.set(name, wr);
-  return image;
+
+  // ...
 }
 ```
 
-But there’s still a problem here: the `Map` still holds on to the `name` strings forever, because those are the keys in the cache. Ideally, those strings would be removed too. The `WeakRef` proposal has a solution for this as well! With the new `FinalizationRegistry` API, we can register a callback to run when the garbage collector zaps a registered object. Such callbacks are known as _finalizers_.
+The downside to this approach is that it is manual memory management. `MovingAvgComponent`, and all other users of the `MovingAvg` class, must remember to call `dispose` or suffer memory leaks. Because the application behavior doesn't depend on the event listener of this diagnostic class, and that the listener is expensive in terms of memory use but not in computation, we really want the listener's lifetime to be logically tied to the `MovingAvg` instance. If its lifetime were tied to the `MovingAvg` instance, then `MovingAvg` could be used like any other JavaScript object whose memory is automatically reclaimed by the garbage collector.
 
-:::note
-**Note:** The finalization callback does not run immediately after garbage-collecting the image object. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
-:::
-
-Here, we register a callback to remove keys from the cache when the image objects are garbage-collected:
+`WeakRef`s make it possible to solve the dilemma by creating a _weak reference_ to the actual event listener, and then wrapping that `WeakRef` in an outer event listener. This way, the garbage collector can clean up the actual event listener and the memory that it holds alive, like the `events` array.
 
 ```js
-const cache = new Map();
-
-const finalizationRegistry = new FinalizationRegistry((name) => {
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
+class MovingAvg {
+  constructor(socket) {
+    this.events = [];
+    const listener = (ev) => { this.events.push(ev); }
+    const weakRef = new WeakRef(listener);
```

Comment:
Ah, doh, this is difficult to get right. I wonder how to keep it alive without making it uncollectible...

Reply:
FWIW since it's part of the WeakRef, at least it'll be guaranteed to stay alive for *1* turn. 😅

Reply:
I think you need to `WeakRef(this)`:

```js
class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.listener = (ev) => { this.events.push(ev); }

    const weakRef = new WeakRef(this);
    const listenerWrapper = (ev) => {
      const deref = weakRef.deref();
      if (deref !== undefined) deref.listener(ev);
    });
    socket.addEventListener("message", listenerWrapper);
  }
}

Reply:
That's trickier to reason through. I plan to update with the following, which is easier to understand and side-steps the non-split function context subtlety:

```js
function makeWeakListener(listener) {
  const weakRef = new WeakRef(listener);
  const wrapper = (ev) => { weakRef.deref()?.(ev); };
  return wrapper;
}

class MovingAvg {
  constructor(socket) {
    this.events = [];
    this.listener = (ev) => { this.events.push(ev); };
    socket.addEventListener("message", makeWeakListener(this.listener));
  }
}
```

As long as you make the wrapper closure outside of the lexical scope of the constructor itself, you can sidestep the trickiness of accidentally entraining variables in the function context.

Reply:
To further clarify why `WeakRef(this)` doesn't work: once `FinalizationRegistry` comes into the picture, we need to pass the wrapper listener to it as a holding value so the finalizer can remove the event listener. That wrapper is strongly referenced as the holdings of the FR. If that wrapper is created inside the lexical scope of the constructor, it'll end up entraining all the other closed over variables. Since `this.listener` is an arrow and closes over `this`, the wrapper thus keeps the `MovingAvg` instance alive. It gets worse: since that `MovingAvg` instance is alive and that instance has a `this.listener` property, that arrow is also kept alive by the FR.

Unfortunately, function contexts is a pretty big footgun that devs will need to understand to use WeakRefs and FRs correctly with functions.